python/cudf_polars/tests/experimental/test_stats.py

```diff
@@ -0,0 +1,353 @@
+                q.select(pl.col("y")),
+            ],
+            how="horizontal",
+        )
+        .group_by(pl.col("y"))
+        .sum()
+        .select(pl.col("x").max(), pl.col("y"))
+        .with_columns((pl.col("x") * pl.col("x")).alias("x2"))
+    )
+    qir2 = Translator(q2._ldf.visit(), engine).translate_ir()
+    stats = collect_base_stats(qir2, ConfigOptions.from_polars_engine(engine))
+    source_info_y = stats.column_stats[qir2]["y"].source_info
+    assert source_info_y.unique_stats("y").fraction == y_unique_fraction
+    assert y_row_count == source_info_y.row_count
+    assert_gpu_result_equal(q2.sort(pl.col("y")).slice(0, 2), engine=engine)
+
+
+@pytest.mark.parametrize("how", ["inner", "left", "right"])
+def test_base_stats_join(how):
+    engine = pl.GPUEngine(
+        raise_on_fail=True,
+        executor="streaming",
+        executor_options={
+            "scheduler": DEFAULT_SCHEDULER,
+            "shuffle_method": "tasks",
+        },
+    )
+    left = pl.LazyFrame(
+        {
+            "x": range(15),
+            "y": [1, 2, 3] * 5,
+            "z": [1.0, 2.0, 3.0, 4.0, 5.0] * 3,
+        }
+    )
+    right = pl.LazyFrame(
+        {
+            "xx": range(9),
+            "y": [2, 4, 3] * 3,
+            "zz": [1, 2, 3] * 3,
+        }
+    )
+    q = left.join(right, on="y", how=how)
+    ir = Translator(q._ldf.visit(), engine).translate_ir()
+    stats = collect_base_stats(ir, ConfigOptions.from_polars_engine(engine))
+
+    ir_column_stats = stats.column_stats[ir]
+    left_count, right_count = 15, 9
+    if how in ("inner", "left"):
+        assert ir_column_stats["x"].source_info.row_count.value == left_count
+        assert ir_column_stats["z"].source_info.row_count.value == left_count
+    if how in ("inner", "right"):
+        assert ir_column_stats["xx"].source_info.row_count.value == right_count
+        assert ir_column_stats["zz"].source_info.row_count.value == right_count
+
+    # TODO: Stats for "y" should depend on join type
+    assert ir_column_stats["y"].source_info.row_count.value == left_count
+
+
+def test_base_stats_union():
+    engine = pl.GPUEngine(
+        raise_on_fail=True,
+        executor="streaming",
+        executor_options={
+            "scheduler": DEFAULT_SCHEDULER,
+            "shuffle_method": "tasks",
+        },
+    )
+    left = pl.LazyFrame(
+        {
+            "x": range(15),
+            "y": [1, 2, 3] * 5,
+            "z": [1.0, 2.0, 3.0, 4.0, 5.0] * 3,
+        }
+    )
+    right = pl.LazyFrame(
+        {
+            "x": range(9),
+            "y": [2, 4, 3] * 3,
+            "z": [1.0, 2.0, 3.0] * 3,
+        }
+    )
+
+    q = pl.concat([left, right])
+    ir = Translator(q._ldf.visit(), engine).translate_ir()
+    stats = collect_base_stats(ir, ConfigOptions.from_polars_engine(engine))
+    column_stats = stats.column_stats[ir]
+
+    # We currently lose source info after a Union.
+    # TODO: We can be "smarter" here.
```

Comment:
An issue to track this would be good, unless you're planning to address that in this PR.